db/table_cache.cc

```diff
@@ -197,4 +201,144 @@ TableCache::GetStatisticValue(
 
 }   // TableCache::GetStatisticValue
 
+
+/**
+ * Riak specific routine to push list of open files to disk
+ */
+void
+TableCache::SaveOpenFileList()
+{
+    Status s;
+    std::string cow_name;
+    WritableFile * cow_file;
+    log::Writer * cow_log;
+
+    cow_name=CowFileName(dbname_);
+    s = env_->NewWritableFile(cow_name, &cow_file, 4*1024L);
+    if (s.ok())
+    {
+        std::string buffer;
+        size_t file_count;
+        cow_log=new log::Writer(cow_file);
+        file_count=doublecache_.WriteCacheObjectWarming(buffer);
+        s = cow_log->AddRecord(buffer);
+        delete cow_log;
+        delete cow_file;
+
+        if (s.ok())
+        {
+            Log(options_->info_log, "Wrote %zd file cache objects for warming.",
+                file_count);
+        }   // if
+        else
+        {
+            Log(options_->info_log, "Error writing cache object file %s (error %s)",
+                cow_name.c_str(), s.ToString().c_str());
+            env_->DeleteFile(cow_name);
+        }   // else
+    }   // if
+    else
+    {
+        Log(options_->info_log, "Unable to create cache object file %s (error %s)",
+            cow_name.c_str(), s.ToString().c_str());
+    }   // else
+
+    return;
+
+}   // TableCache::SaveOpenFiles
+
+/**
+ * Riak specific routine to read list of previously open files
+ *  and preload them into the table cache
+ */
+void
+TableCache::PreloadTableCache()
+{
+    struct LogReporter : public log::Reader::Reporter {
+        Env* env;
+        Logger* info_log;
+        const char* fname;
+        Status* status;  // NULL if options_.paranoid_checks==false
+        virtual void Corruption(size_t bytes, const Status& s) {
+            Log(info_log, "%s%s: dropping %d bytes; %s",
+                (this->status == NULL ? "(ignoring error) " : ""),
+                fname, static_cast<int>(bytes), s.ToString().c_str());
+            if (this->status != NULL && this->status->ok()) *this->status = s;
+        }
+    };
+
+    Status s;
+    std::string cow_name;
+    SequentialFile * cow_file;
+    log::Reader * cow_log;
+    size_t obj_count(0);
+
+    cow_name=CowFileName(dbname_);
+    s = env_->NewSequentialFile(cow_name, &cow_file);
+    if (s.ok())
+    {
+        // Create the log reader.
+        LogReporter reporter;
+        reporter.env = env_;
+        reporter.info_log = options_->info_log;
+        reporter.fname = cow_name.c_str();
+        reporter.status = &s;
+
+        std::string buffer;
+        Slice record;
+
+        cow_log=new log::Reader(cow_file, &reporter, true, 0);
+
+        while (cow_log->ReadRecord(&record, &buffer) && s.ok())
+        {
+            Slice input = record;
+            uint32_t tag, level;
+            uint64_t file_no, file_size;
+            Cache::Handle * handle;
+            Status s2;
+
+            while (GetVarint32(&input, &tag))
```

Comment:
I recommend this set of GetVarintXx() calls be moved to cache2.cc as a companion to the SharedLRUCache2::WriteFileCacheObjectWarming() method, which writes the serialized record. The serialization/deserialization logic should be kept together, if at all possible.

You could declare a structure whose data members are level, file_no, and file_size, and pass a reference to a std::list<> of this structure that the new method would populate from the input Slice object.

Then this method would change from iterating over the GetVarintXx() calls to iterating over the std::list<>, calling FindTable() for each entry in the list.